### PR TITLE
kubeadm: turn api into a real apigroup

### PR DESCRIPTION
**Pull request by mikedanese**
@kubernetes/sig-cluster-lifecycle

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34147)

### CODE REVIEW

**Comment by justinsb** [comment]
+1 on the idea @mikedanese  (sadly I don't know enough to review the code).  And thanks for providing an example on how to use the k8s api machinery.  (Did you find a nice doc on how to do this BTW?)

**Review by krousey — COMMENTED** [review]
I'm fine with the code change, but are we sure we want to use the API machinery for what is essentially a versioned config file? cc @kubernetes/sig-api-machinery
Things I usually associate with an API group that don't apply here:
- Swagger definitions
- API discovery
- Storage interfaces
- Client
- REST endpoints
- alternate encodings

If we do use the API, would this work without ugorji codec?

File: cmd/kubeadm/app/api/install/install.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (lavalamp) [inline cmd/kubeadm/app/api/install/install.go]: please make the path to this file (and other api machinery files) be `.../apis/<your group name>/...`, not `/api`

File: cmd/kubeadm/app/api/install/install.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		&announced.GroupMetaFactoryArgs{
+			GroupName:                  api.GroupName,
+			VersionPreferenceOrder:     []string{v1alpha1.SchemeGroupVersion.Version},
+			ImportPrefix:               "k8s.io/kubernetes/cmd/kubeadm/api",
+			AddInternalObjectsToScheme: api.AddToScheme,
+		},
+		announced.VersionToSchemeFunc{
+			v1alpha1.SchemeGroupVersion.Version: v1alpha1.AddToScheme,
+		},
+	).Announce().RegisterAndEnable(); err != nil {
```
Comment (lavalamp) [inline cmd/kubeadm/app/api/install/install.go]: Do you actually need to register this group?

**Comment by errordeveloper** [comment]
> I'm fine with the code change, but are we sure we want to use the API machinery for what is essentially a versioned config file?

Yes, good question, I wonder about this also.

**Comment by errordeveloper** [comment]
@mikedanese thanks a lot for doing this! As far as I understand, this is not in use yet, so are you planning to make use of it in a follow-up PR or you will add it here?

File: cmd/kubeadm/app/api/install/install.go (hunk 1 above)
Comment (mikedanese) [inline cmd/kubeadm/app/api/install/install.go]: What does registering an API group provide? I only need versioning, defaulting and documentation.

**Comment by mikedanese** [comment]
> are you planning to make use of it in a follow-up PR or you will add it here?

End goal is to make this deserialize the internal config representation from a file passed in from the command line for the "configurable mode" experience. With this we can hopefully avoid diluting the streamlined command line experience.

**Comment by mikedanese** [comment]
> I'm fine with the code change, but are we sure we want to use the API machinery for what is essentially a versioned config file?

I feel like this ship flew with componentconfig. Aren't these requirements (versioning, defaulting and documentation) the some of the core use cases for API groups?

**Comment by lavalamp** [comment]
It will make it an option in the --runtime-config flag, which I guess
doesn't matter if you are working in a separate binary.

**Comment by mikedanese** [comment]
Ok unregistered the group.

**Comment by mikedanese** [comment]
What about roundtrip testing?

**Comment by dgoodwin** [comment]
I think previous commit 56ea178e7c77ab88b949107c420bc58d593143b3 may have broken something, the controller manager is unable to start up with a plain "kubeadm init master" now:

```
(root@centos1 ~) $ docker logs 589b79ef9fd9
E1007 18:44:49.091488       1 event.go:258] Could not construct reference to: '&api.Endpoints{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"kube-controller-manager", GenerateName:"", Namespace:"kube-system", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Subsets:[]api.EndpointSubset(nil)}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Normal' '%v became leader' 'centos1.aos.example.com'
I1007 18:44:49.092504       1 leaderelection.go:214] sucessfully acquired lease kube-system/kube-controller-manager
I1007 18:44:49.098480       1 plugins.go:71] No cloud provider specified.
W1007 18:44:49.098495       1 controllermanager.go:232] Unsuccessful parsing of cluster CIDR : invalid CIDR address: 
W1007 18:44:49.098503       1 controllermanager.go:236] Unsuccessful parsing of service CIDR : invalid CIDR address: 
I1007 18:44:49.099198       1 replication_controller.go:220] Starting RC Manager
I1007 18:44:49.099747       1 nodecontroller.go:194] Sending events to api server.
F1007 18:44:49.099815       1 nodecontroller.go:206] NodeController: Must specify clusterCIDR if allocateNodeCIDRs == true.
```

**Comment by mikedanese** [comment]
@dgoodwin before in `kubeadm init master --pod-network-cidr` was `--pod-network-cidr` a required flag? Should it be required? I remember @errordeveloper saying it shouldn't be. If it shouldn't be then https://github.com/kubernetes/kubernetes/pull/34352 might fix your issue.

**Review by krousey — APPROVED** [review]

**Comment by errordeveloper** [comment]
@mikedanese @dgoodwin `--pod-network-cidr` should be optional.

**Comment by mikedanese** [comment]
@krousey is that an lgtm? :)

**Comment by luxas** [comment]
@errordeveloper Yes, that was probably a rebase conflict or something from some previous refactoring PR (after our "stable" v1.4 release)

@mikedanese I think that's a LGTM, yes

@errordeveloper Any comments on this one?

**Comment by errordeveloper** [comment]
@mikedanese:
> End goal is to make this deserialize the internal config representation from a file passed in from the command line for the "configurable mode" experience. With this we can hopefully avoid diluting the streamlined command line experience.

Yes, I'm familiar with the goal already 👍 :)

I am just wondering how far does this change set get us, I cannot see any flags being added here explicitly, I'd expect there to be at least one more change set to plumb it through, yet I am not aware if there is any magic that would infer any automatic flags for passing the file in...

**Comment by krousey** [comment]
Yes

**Pull request merged**